security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -247,27 +258,38 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
-	// In the current implementation, the file update events are not handled and if
-	// the files are updated, a user may restart Envoy to pick up the updated files.
-	// TODO (lei-tang): if updating files are supported, add a file watcher for
-	// the files under the well known path.
 	sdsFromFile := false
 	var err error
+	var ns *model.SecretItem
 
-	if connKey.ResourceName == RootCertReqResourceName && sc.rootCertificateExist(sc.existingRootCertFile) {
+	switch {
+	// Default root certificate.
+	case connKey.ResourceName == RootCertReqResourceName && sc.rootCertificateExist(sc.existingRootCertFile):
 		sdsFromFile = true
 		ns, err = sc.generateRootCertFromExistingFile(sc.existingRootCertFile, token, connKey)
-	} else if cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName); ok && cfg.CaCertificatePath != "" {
-		// Based on the resource name, we need to read this secret from a file encoded in the resource name
-		sdsFromFile = true
-		ns, err = sc.generateRootCertFromExistingFile(cfg.CaCertificatePath, token, connKey)
-	} else if connKey.ResourceName == WorkloadKeyCertResourceName && sc.keyCertificateExist(sc.existingCertChainFile, sc.existingKeyFile) {
+		sc.addFileWatcher(sc.existingRootCertFile, connKey)
+	// Defautl workload certificate.
+	case connKey.ResourceName == WorkloadKeyCertResourceName && sc.keyCertificateExist(sc.existingCertChainFile, sc.existingKeyFile):
 		sdsFromFile = true
 		ns, err = sc.generateKeyCertFromExistingFiles(sc.existingCertChainFile, sc.existingKeyFile, token, connKey)
-	} else if cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName); ok && cfg.CertificatePath != "" && cfg.PrivateKeyPath != "" {
-		// Based on the resource name, we need to read this secret from a file encoded in the resource name
-		sdsFromFile = true
-		ns, err = sc.generateKeyCertFromExistingFiles(cfg.CertificatePath, cfg.PrivateKeyPath, token, connKey)
+		sc.addFileWatcher(sc.existingCertChainFile, connKey)
+		sc.addFileWatcher(sc.existingKeyFile, connKey)
+	default:
+		// Check if the resource name refers to a file mounted certificate.
+		// Currently used in destination rules and server certs (via metadata).
+		// Based on the resource name, we need to read the secret from a file encoded in the resource name.
+		cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName)
+		switch {
+		case ok && cfg.IsRootCertificate() && sc.rootCertificateExist(cfg.CaCertificatePath):
+			sdsFromFile = true
+			ns, err = sc.generateRootCertFromExistingFile(cfg.CaCertificatePath, token, connKey)
+			sc.addFileWatcher(cfg.CaCertificatePath, connKey)
+		case ok && cfg.IsKeyCertificate() && sc.keyCertificateExist(cfg.CertificatePath, cfg.PrivateKeyPath):
+			sdsFromFile = true
+			ns, err = sc.generateKeyCertFromExistingFiles(cfg.CertificatePath, cfg.PrivateKeyPath, token, connKey)
+			sc.addFileWatcher(cfg.CertificatePath, connKey)
+			sc.addFileWatcher(cfg.PrivateKeyPath, connKey)
```

Comment:
Can you just watch the cert not the key? A key can't change without changing the cert.
I'm asking because you may receive two update events for a key&cert pair update, and cause two push events. We don't want that to happen.

Reply:
Sure. That was my original implementation https://github.com/istio/istio/pull/23417#discussion_r418107281 - But added the other one later. I can go back to previous implementation